Added initial docs about what the package does

astrofrog opened the pull request:
I thought it was about time we documented what astropy-helpers actually does. This is pretty minimal, but I thought it'd be better than nothing.

mhvk commented:
Thanks for adding this! One possible addition that would be great -- but don't let it stand in the way of merging this! -- would be to tell also *where* these additions are done (i.e., link to the relevant files). In this respect, would it be crazy to run sphinx here too? Most functions seem very nicely documented. (On the other hand, if we're thinking of getting rid of this, then it would be a waste of time...)

saimn commented:
Maybe mention also `setup_package.py` files ?

cdeil commented:
@astrofrog - Thank you!

For me as affiliated package maintainer, the most interesting information would be how to update `astropy-helpers` and the related files we have inherited from the package-template, especially `setup.py` and `_astropy_init.py`.

So maybe add a mention of `_astropy_init.py`?

Also, if the bot is still running that will send PRs to do the updates, it should be mentioned as well?

There are other related docs pages, but they also don't give a clear path how to update:
http://docs.astropy.org/en/stable/development/astropy-package-template.html
http://docs.astropy.org/projects/package-template/en/latest/updating.html

Would one ever use `astropy-helpers` without the `_astropy_init.py` that comes from the package-template? Would you update the `astropy-helpers` always together with these few extra files that live outside the `astropy-helpers` folder or always in sync?

Basically: somewhere there should be a description of when / how to apply updates. I think that's the hard and time-consuming part for maintainers. Running cookie-cutter once is probably the easy part and already well-documented. So a better understanding / description how the few extra files from the package template relate to astropy-helpers (`_astropy_init.py`, `ah_bootstrap.py` and `setup.py` and `setup.cfg` mainly, I think) and how to update those would be most valuable for me.

astrofrog commented:
@cdeil - the ``_astropy_init.py`` file is mostly unrelated to astropy-helpers. I think the best place to put the instructions you mention would actually be the package-template docs: http://docs.astropy.org/projects/package-template/en/latest

mhvk commented:
I fear I have to second @cdeil here - The documentation for getting started with an affiliated package is really good but I've found keeping up to date tricky. Partially it is of course that it needs to be done only rarely, so all I remember is that I went by trial and error. And partially it is that I started before cookiecutter so have to do things manually. But partially also it is very difficult to know whether changes affects one's package.

One concrete recommendation is to provide links to changelogs - I know I found a summary of changes before (I think just on github), but the links above to not point to them.

Anyway, this is getting really a bit off-topic here!  And I fear I understand it too poorly to even help document it better -- which may be another good reason to try to minimize the number of moving parts.

cdeil commented on README.rst:
```diff
@@ -23,22 +24,117 @@ implementation are in the accepted
+
+* ``-o``: open the documentation in a browser if a build finishes successfully.
+
+In addition, build_docs will automatically download and temporarily install sphinx-astropy
```
What does "temporarily install sphinx-astropy" mean?
Is it installed and removed again? In which environment / site-packages?
Does this always happen or only on first run?

cdeil commented on README.rst:
```diff
@@ -23,22 +24,117 @@ implementation are in the accepted
+
+We redefine ``sdist`` to use the version from distutils rather than from
+setuptools, as the setuptools version requires duplication of information
+in ``MANIFEST.in``.
```
Is this customisation of `sdist` something that could maybe be removed completely?
I find `MANIFEST.in` confusing and don't understand what the good practices are concerning what include and exclude lines to put, i.e. whether to white or blacklist files, from the PYthon documentation.

So I can see why a modifcation was introduced here as an attempt to make it better.

But probably making it behave differently in Python packages that use astropy-helpers and "normal" Python packages that use setuptools isn't a good idea? It just creates more confusion overall?

bsipocz replied:
see #419 for the discussion or removal of the custom `sdist`.

astrofrog commented:
Thanks all for your feedback!

@cdeil: I've added instructions on manually installing and updating astropy-helpers in a package. I've not added anything about _astropy_init.py since that's unrelated and package-template specific. I've also tried to explain better what temporary installation means.

@mhvk: documenting how to update other parts of packages using the package template is an issue for the package-template repo

@saimn: I've added a description of the ``setup_package.py`` files. This was actually present in the astropy docs hidden away in the developer docs so I stole some text from there:

http://docs.astropy.org/en/latest/development/building.html#customizing-setup-build-for-subpackages

I think it makes more sense for it to live here, so once this PR is merged, we can just add a link to this README from the astropy docs.

I think this is now a reasonably complete record of most things that can be done with astropy-helpers. Yes, it could live in sphinx docs, but let's see first what can be removed/simplified.

cdeil commented:
@astrofrog - Thank you very much. Having this description / reference is super useful. Maybe just merge this in now?

astrofrog commented:
@cdeil - I agree, I'll just wait on approvals from @eteq and @bsipocz as they are officially maintainers of the helpers.

bsipocz commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+To instead update astropy-helpers manually, go inside the submodule and do::
+
+    cd astropy_helpers
+    git fetch origin
```
Most cases this is enough, but sometimes the tags are not coming through automatically with this line and I needed to add `--tags` to the fetch (e.g. it's definitely the case for the tag v3.1rc1). I don't understand the reason maybe something I messed up when releasing.

eteq replied:
I've seen the same thing and not fully understood why... but I think it might have to do with the the signing?  Or possibly *editing* of tags?

bsipocz replied:
My unfounded suspicion is that it happens when the tag is at the top of a branch? I'll try to remember to double check whether this is the case once the 3.1.x branch is updated.

bsipocz replied:
just to confirm this is almost certainly an issue I've created by only pushing the tag back to github, but not the updated branch. Since the latter has been done I cannot reproduce it, the latest tag is available without doing `--tags`.

bsipocz commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+Another piece of functionality we provide in astropy-helpers is the ability
+to generate a ``packagename.version`` file that includes functions that
+automatically set the version string for developer versions, to e.g.
+``3.2.dev22213`` so that each developer version has a unique number. To use this,
```
This may be a bug or a feature, but people working on separate branches will share version numbers for different commits, so this uniqueness cannot be relied on on the package level.

astrofrog replied:
I've tried to clarify this.

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+
+The easiest way to get set up with astropy-helpers in a new package is to use
+the `package-template <http://docs.astropy.org/projects/package-template>`_
+that we provide. However, we now go through the steps of adding astropy-helpers
```
```suggestion
that we provide. This template is specifically designed for use with the helpers,
so using it avoids some of the tedium of setting up the heleprs.

However, we now go through the steps of adding astropy-helpers
```

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+changes::
+
+    git add astropy_helpers ah_bootstrap.py setup.py
+    git commit ...
```
```suggestion
    git commit -m "started using astropy-helpers"
```

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+------------------------
+
+If you would like to automatically get pull requests to update astropy-helpers,
+then see the instructions `here
```
This link doesn't seem to point to any actual instructions. What file actually has the instructions?  I suggest linking there instead, along with maybe some words like "If you would like the Astropy team to automatically generate pull requests..." - that is, make it clear that "automatically" really means "when there's a reason to do so" (assuming that's true, which I think it is, right?)

astrofrog replied:
Fixed

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+    git add astropy_helpers ah_bootstrap.py
+    git commit ...
+
+What does astropy-helpers provide?
```
Maybe this section should come *first*?  I.e., before the "how to get it"?

astrofrog replied:
I've re-ordered it

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+This command will automatically build the package, install it to a temporary
+directory, and run the tests using `pytest <http://pytest.org/>`_ on this
+installed version. Note that the bulk of this command is actually defined
+in ``astropy.tests.command.AstropyTest`` - here we simply define the custom
```
Maybe say *why* this is here?  (I.e. maybe:)
```suggestion
in ``astropy.tests.command.AstropyTest``, because that allows the test 
machinery to operate outside a setuptools command. This, here we 
simply define the custom
```

astrofrog replied:
Added

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+**python setup.py build_docs**
+
+This command will automatically build the package, then run sphinx to build
+the documentation. Sphinx itself provides a custom setuptools command, which we
```
```suggestion
the documentation. This makes development much easier because it ensures
sphinx extensions that use the package's  code to make documentation are 
actually using the in-development version of the code. Sphinx itself
provides a custom setuptools command, which we
```

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+  installed/available before the package is installed, which means that one
+  needs to install Numpy in a separate installation step.
+
+* We detect broken compilers and replace them with other compilers on-the-fly
```
(reminder: remove this if #421 goes in, or remove it *in* #421 if this goes in first)

astrofrog replied:
I think this will go in first.

eteq commented on README.rst:
```diff
@@ -23,22 +30,304 @@ implementation are in the accepted
+
+What does astropy-helpers provide?
+----------------------------------
+
```
Feel free to edit this, but I think a top level "what does it do big-picture" would help a bit.

```suggestion

Astropy-helpers' big-picture purpose is to provide customization to Python's
packaging infrastructure process in ways that the Astropy Project has found to
help simplifying the developint and releasing packages. This is primarily
build around ``setup.py`` commands, as outlined below, as well as code to help
manage version numbers and better control the build process of larger packages.

```

astrofrog commented on CHANGES.rst:
```diff
@@ -4,7 +4,7 @@ astropy-helpers Changelog
 ----------------
 
-
+- Added extensive documentation about astropy-helpers to the README.rst file. [#416]
```
I've moved this to 3.1. I'm reluctant to move to 2.0.x since indeed it won't actually all be correct there.

astrofrog commented:
> I'm a bit worried about making sure this stays up-to-date... But eh, it's better to have it now and try to keep it coherent than not having it, given the complexity of this process.

One could say that about the entire astropy developer docs ;)

I've now implemented all the comments - since @eteq and @bsipocz approved this, I'll merge once the CI passes.
